File: tasks/configuration-merge/python/environment/src/config_merge.py
import json
import sys


def validate(value):
    keys = {"defaults", "file", "env", "cli"}
    if not isinstance(value, dict) or set(value) != keys:
        raise ValueError("invalid input")
    if any(not isinstance(value[key], dict) for key in keys):
        raise ValueError("invalid layer")


def merge(value):
    validate(value)
    result = {}
    for key in ("defaults", "file", "env", "cli"):
        result.update(value[key])
    return result


try:
    print(json.dumps(merge(json.load(sys.stdin)), separators=(",", ":")))
except Exception as error:
    print(str(error), file=sys.stderr)
    raise SystemExit(1)

File: tasks/configuration-merge/javascript/environment/src/config_merge.js
const fs = require("node:fs");

function validate(input) {
  const keys = ["defaults", "file", "env", "cli"];
  if (!input || typeof input !== "object" || Array.isArray(input)) throw new Error("invalid input");
  if (Object.keys(input).sort().join(",") !== keys.slice().sort().join(",")) throw new Error("invalid layers");
  for (const key of keys) {
    if (!input[key] || typeof input[key] !== "object" || Array.isArray(input[key])) throw new Error("invalid layer");
  }
}

function merge(input) {
  validate(input);
  return Object.assign({}, input.defaults, input.file, input.env, input.cli);
}

try {
  const input = JSON.parse(fs.readFileSync(0, "utf8"));
  process.stdout.write(JSON.stringify(merge(input)) + "\n");
} catch (error) {
  process.stderr.write(String(error.message || error) + "\n");
  process.exitCode = 1;
}
